ocaml/xapi/xapi_session.ml

```diff
@@ -27,6 +27,228 @@ open Client
+
+    let record_unknown () =
+      Mutex.execute m (fun () ->
+          let ctr = !unknown_ctr + 1 in
+          unknown_ctr := ctr ;
+          ctr)
+
+    type key = client
+
+    type value = {num_failed_attempts: int; last_failed_attempt: Date.iso8601}
+
+    type table_t = (key, value) Hashtbl.t
+
+    let table = Hashtbl.create 10
+
+    let record_client k ~now =
+      Mutex.execute m (fun () ->
+          match Hashtbl.find_opt table k with
+          | None ->
+              Hashtbl.add table k
+                {num_failed_attempts= 1; last_failed_attempt= now} ;
+              1
+          | Some ({num_failed_attempts} : value) ->
+              let num_failed_attempts = num_failed_attempts + 1 in
+              Hashtbl.replace table k
+                {num_failed_attempts; last_failed_attempt= now} ;
+              num_failed_attempts)
+
+    let get () =
+      let reset () =
+        Hashtbl.reset table ;
+        unknown_ctr := 0
+      in
+      Mutex.execute m (fun () ->
+          let unknown_client_failed_attempts = !unknown_ctr in
+          if Hashtbl.length table = 0 && unknown_client_failed_attempts = 0 then
+            None
+          else
+            let num_known_client_failed_attempts, top_3_worst_clients =
+              Hashtbl.fold
+                (fun client {num_failed_attempts; last_failed_attempt}
+                     (ctr, top_three) ->
+                  ( ctr + num_failed_attempts
+                  , TopThree.maybe_add top_three
+                      {client; num_failed_attempts; last_failed_attempt} ))
+                table (0, [])
+            in
+            reset () ;
+            Some
+              {
+                total_num_failed_attempts=
+                  num_known_client_failed_attempts
+                  + unknown_client_failed_attempts
+              ; top_3_worst_clients= List.rev top_3_worst_clients
+              ; unknown_client_failed_attempts
+              })
+  end
+
+  let get_stats_string () = Stats.get () |> Option.map string_of_stats
+
+  let on_fail ~__context ~now ~uname ~originator ~record =
+    try
+      match client_of_info ~__context ~uname ~originator with
+      | None ->
+          let total_unknown_login_failures = Stats.record_unknown () in
+          warn "login failure from unknown client, total= %i"
+            total_unknown_login_failures
+      | Some client ->
+          let num_failed_attempts = Stats.record_client client ~now in
+          info "failed login attempt #%i by client: %s" num_failed_attempts
+            (string_of_client client)
+    with e ->
+      (* we don't expect this function to fail, but if it does we don't want to block callers *)
+      error "AuthFail.on_fail_with_uname: unexpected error: '%s'"
+        (Printexc.to_string e)
+end
+
+let _record_login_failure ~__context ~now ~uname ~originator ~record f =
+  try f ()
+  with e ->
+    (* naively assume any exception corresponds to an authentication failure *)
```

Comment:
How reliable is this? Don't we have specific exception to signal authentication failure?

Reply:
> How reliable is this?
It depends how you define reliable

false positive = FP = we logged a failed login request as an auth failure when it wasn't really an auth failure
false negative = FN = an login failed because of invalid credentials and we didn't record it

The current implementation has the potential to cause FPs, but will never result in a FN. If we catch only on `Auth_signature.Auth_failure` and the code behaves exactly as we expect, then we'll never get FNs or FPs :+1: However my confidence of that isn't particularly high after looking at `xapi_session.ml:login_with_password`

In practice it probably doesn't matter which method we choose

Reply:
Could the code in the logins be changed to use `Auth_failure` when appropiate? I would prefer to minimize false positives to have minimum noise if there are no auth failures. Customers might get worried at first and call support,or if it becomes too noisy they will tend to ignore this kind of alerts.

Reply:
To be clear, you are talking about looking at the implementations of the login functions (e.g. `login_with_password`), and making sure they from `Auth_failure` when we expect? I will look into this

Reply:
That's correct. Makes those function throw `Auth_failure` so we can rely on it for logging / alerting.